### PR TITLE
Fix/prefix cache stream mismatch

### LINKED ISSUE #800: [Bug]: Prefix cache hit causes "Detected MLX stream/thread mismatch on worker step"
### Environment

- Chip: M4 Max
- RAM: 36GB
- macOS version: 26.6.2
- Python version: 3.13.15

Name: vllm-mlx
Version: 0.4.1
Name: mlx
Version: 0.32.2
Name: mlx-lm
Version: 0.31.3
Name: mlx-vlm
Version: 0.7.1
Name: mlx-audio
Version: 0.5.4
Name: mlx-embeddings
Version: 0.1.0

### Model and quantization

Qwen/Qwen2.5-7B-Instruct (reproduced on any model with --continuous-batching and prefix caching)

### Reproduction

```shell
1. Clean cache (or don't, I do it just to be more clear)

rm -rf ~/.cache/vllm-mlx/prefix_cache/*

2. vllm-mlx serve Qwen/Qwen2.5-7B-Instruct --continuous-batching --host 0.0.0.0 --port 8000

3. Make 2 requests 

curl -s -X POST http://localhost:8000/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{
    "model": "Qwen/Qwen2.5-7B-Instruct",
    "messages": [
      {
        "role": "system",
        "content": "You are a helpful, respectful and honest assistant. Please carefully analyze the following background information and summarize the core principles. Background: Apple Silicon features a unified memory architecture where CPU, GPU, and Neural Engine share the same memory pool. The MLX framework is designed for machine learning research on Apple Silicon, taking advantage of unified memory and lazy evaluation. When serving models asynchronously in production, concurrency models must be designed with care. An event loop handling HTTP requests in FastAPI operates on the main thread, while model generation typically executes on a background worker thread. If MLX arrays are evaluated across different threads or streams without synchronization, stream mismatch issues can arise."
      },
      {
        "role": "user",
        "content": "What are the key points?"
      }
    ],
    "max_tokens": 10
  }'


curl -s -X POST http://localhost:8000/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{
    "model": "Qwen/Qwen2.5-7B-Instruct",
    "messages": [
      {
        "role": "system",
        "content": "You are a helpful, respectful and honest assistant. Please carefully analyze the following background information and summarize the core principles. Background: Apple Silicon features a unified memory architecture where CPU, GPU, and Neural Engine share the same memory pool. The MLX framework is designed for machine learning research on Apple Silicon, taking advantage of unified memory and lazy evaluation. When serving models asynchronously in production, concurrency models must be designed with care. An event loop handling HTTP requests in FastAPI operates on the main thread, while model generation typically executes on a background worker thread. If MLX arrays are evaluated across different threads or streams without synchronization, stream mismatch issues can arise."
      },
      {
        "role": "user",
        "content": "Give me a short 5-word summary please."
      }
    ],
    "max_tokens": 10
  }'
```

### Expected behavior

I expect the background worker to handle the MLX, not the MainThread

### Actual behavior

```shell
I see this in logs, because MainThread used mlx functions (created tensors) and we cannot pass the work to the worker

WARNING:vllm_mlx.engine_core:Detected MLX stream/thread mismatch on worker step; switched this engine to model-thread stepping
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by riiji**
fixes #800 

## Summary

This PR introduces `MLXExecutor` — a centralized thread-routing abstraction for MLX operations — and fixes critical stream mismatch errors, throughput collapse, and infinite runaway loops during continuous batching:

1. **Centralized MLX Thread Ownership (`vllm_mlx/mlx_executor.py`)**:
   - MLX arrays and streams are thread-local; evaluating tensors created on thread A from thread B throws `There is no Stream(gpu, N) in current thread`.
   - `MLXExecutor` owns the single generation worker pool and ensures all operations touching MLX (model forward pass, scheduler stepping, prefix cache load/save, and cache clear) execute on the owner thread.
   - **Zero dispatch overhead**: when called from the owner thread, `run()` and `arun()` execute inline with 0 overhead.

2. **Fix Stream Mismatch on Cache Persistence**:
   - Previously, `load_cache_from_disk()` and `save_cache_to_disk()` were invoked on the `MainThread` (or via `asyncio.to_thread` on arbitrary worker threads) during server lifespan startup/shutdown.
   - This created KV cache tensors on `MainThread`'s default stream, causing stream mismatch crashes when the generation worker attempted to evaluate them.
   - Now, all cache persistence methods in `EngineCore` and `BatchedEngine` route through `MLXExecutor.run()`.

3. **Eliminate Catastrophic MainThread Fallback in `EngineCore`**:
   - Previously, if a stream error occurred in `_engine_loop`, `use_worker_thread = False` permanently switched all subsequent `scheduler.step()` calls onto the asyncio Event Loop (`MainThread`).
   - This starved networking and JSON serialization, collapsing throughput from 80+ tok/s to <2 tok/s.
   - Replaced this fallback with in-worker stream rebinding and request rescheduling, keeping MLX operations strictly off the Event Loop.

4. **Add Runaway-Loop Watchdog**:
   - If requests were stranded in `scheduler.running` while `batch_generator.next()` returned empty results, the engine loop previously spun at ~50k steps/second until process OOM-kill.
   - Added a watchdog in `_engine_loop` that aborts stranded requests with `finish_reason="error"` if consecutive empty steps exceed the threshold.

## Type of change

- Bug fix
- Performance
- Refactor or internal cleanup

## Surface touched

- Scheduler / batching
- KV cache / prefix cache
- OpenAI API endpoints
- CLI / serve command

### 1. Automated unit and regression tests:
```bash
pytest tests/test_batched_engine_owner_thread.py \
       tests/test_engine_core_thread_streams.py \
       tests/test_engine_core_idle_polling.py \
       tests/test_mlx_executor.py \
       tests/test_engine_core_watchdog.py \
       tests/test_batched_engine_cache_dispatch.py \
       tests/test_server_cache_controls.py -v
```

### 2. Live server benchmark:
```bash
# Terminal 1: Start server
vllm-mlx serve mlx-community/Qwen2.5-7B-Instruct-4bit --continuous-batching --port 8000

# Terminal 2: Run benchmark sweep
vllm-mlx bench-serve --url http://127.0.0.1:8000 --validate false
```

## Test results

All 25 tests pass cleanly:
```
tests/test_batched_engine_owner_thread.py::test_generation_worker_is_one_reused_thread PASSED [  4%]
tests/test_batched_engine_owner_thread.py::test_worker_thread_is_named_for_the_engine_loop PASSED [  8%]
tests/test_batched_engine_owner_thread.py::test_model_load_and_stepping_share_one_thread PASSED [ 12%]
tests/test_batched_engine_owner_thread.py::test_mllm_load_and_step_share_the_event_loop_thread PASSED [ 16%]
tests/test_batched_engine_owner_thread.py::test_residency_manager_honours_the_per_path_load_thread PASSED [ 20%]
tests/test_batched_engine_owner_thread.py::test_supplied_worker_outlives_the_engine_loop PASSED [ 24%]
tests/test_batched_engine_owner_thread.py::test_loop_still_cleans_up_a_worker_it_created PASSED [ 28%]
tests/test_batched_engine_owner_thread.py::test_stop_retires_the_generation_worker PASSED [ 32%]
tests/test_batched_engine_owner_thread.py::test_async_engine_core_passes_the_worker_through PASSED [ 36%]
tests/test_batched_engine_owner_thread.py::test_add_request_runs_on_generation_worker PASSED [ 40%]
tests/test_engine_core_thread_streams.py::test_engine_core_runs_all_scheduler_steps_on_one_worker_thread PASSED [ 44%]
tests/test_engine_core_thread_streams.py::test_mllm_scheduler_runs_steps_on_model_load_thread PASSED [ 48%]
tests/test_engine_core_idle_polling.py::test_engine_loop_uses_idle_interval_when_scheduler_is_empty PASSED [ 52%]
tests/test_engine_core_idle_polling.py::test_engine_loop_wakes_promptly_for_request_added_while_parked PASSED [ 56%]
tests/test_engine_core_idle_polling.py::test_engine_config_keeps_existing_positional_argument_mapping PASSED [ 60%]
tests/test_mlx_executor.py::test_executor_runs_on_dedicated_thread PASSED [ 64%]
tests/test_mlx_executor.py::test_executor_inline_detection_on_owner_thread PASSED [ 68%]
tests/test_mlx_executor.py::test_executor_arun_from_event_loop PASSED    [ 72%]
tests/test_mlx_executor.py::test_executor_arun_inline_async PASSED       [ 76%]
tests/test_mlx_executor.py::test_executor_bind_streams_runs_on_worker PASSED [ 80%]
tests/test_mlx_executor.py::test_executor_unowned_worker_not_shut_down PASSED [ 84%]
tests/test_engine_core_watchdog.py::test_watchdog_aborts_stranded_requests_after_max_empty_steps PASSED [ 88%]
tests/test_batched_engine_cache_dispatch.py::test_cache_operations_run_on_mlx_executor_thread PASSED [ 92%]
tests/test_server_cache_controls.py::test_cache_stats_includes_engine_cache PASSED [ 96%]
tests/test_server_cache_controls.py::test_clear_cache_clears_engine_managed_runtime_caches PASSED [100%]

============================== 25 passed in 4.05s ==============================
```

## Performance impact

- **Hardware**: Apple M4 Max, 36GB Unified Memory, macOS 26.6.2
- **Model**: `mlx-community/Qwen2.5-7B-Instruct-4bit`
- **Benchmark Command**: `vllm-mlx bench-serve --url http://127.0.0.1:8000 --validate false`

### PR Branch (`fix/prefix-cache-stream-mismatch`):
| prompt_set | concurrency | prompt_tokens | ttft_ms | tpot_ms | gen_tps | prompt_tps | e2e_latency_ms |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| short | 1 | 43 | 119.0 | 12.9 | 77.7 | 361.3 | 3416.3 |
| short | 1 | 43 | 125.6 | 12.5 | 80.4 | 342.3 | 3311.7 |
| short | 1 | 43 | 128.0 | 12.5 | 80.4 | 336.0 | 3312.5 |
| short | 4 | 43 | 476.8 | 38.5 | 26.1 | 92.5 | 10301.1 |
| short | 4 | 43 | 449.2 | 38.5 | 26.1 | 98.1 | 10265.9 |
| short | 4 | 43 | 456.9 | 38.3 | 26.2 | 96.5 | 10226.3 |
| medium | 1 | 568 | 825.6 | 12.9 | 77.9 | 688.0 | 4111.4 |
| medium | 1 | 568 | 830.6 | 12.9 | 78.0 | 683.9 | 4113.5 |
| medium | 1 | 568 | 829.5 | 12.9 | 77.8 | 684.7 | 4122.6 |
| medium | 4 | 568 | 3967.8 | 39.8 | 25.2 | 134.8 | 14113.6 |
| medium | 4 | 568 | 3808.1 | 39.7 | 25.3 | 140.5 | 13933.3 |
| medium | 4 | 568 | 3787.2 | 40.7 | 24.7 | 141.3 | 14164.0 |
| long | 1 | 2513 | 4140.0 | 14.0 | 71.7 | 607.0 | 7712.0 |
| long | 1 | 2513 | 4339.3 | 14.4 | 69.6 | 579.1 | 8019.7 |
| long | 1 | 2513 | 4495.8 | 14.9 | 67.4 | 559.0 | 8296.1 |
| long | 4 | 2513 | 15281.0 | 26.8 | 37.5 | 119.7 | 22117.2 |
| long | 4 | 2513 | 14963.6 | 25.9 | 38.8 | 122.2 | 21558.3 |
| long | 4 | 2513 | 14365.2 | 24.9 | 40.3 | 127.3 | 20714.4 |

### Baseline (`main`):
| prompt_set | concurrency | prompt_tokens | ttft_ms | tpot_ms | gen_tps | prompt_tps | e2e_latency_ms |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| short | 1 | 43 | 131.2 | 12.7 | 79.2 | 327.6 | 3363.9 |
| short | 1 | 43 | 123.4 | 12.7 | 78.8 | 348.4 | 3373.2 |
| short | 1 | 43 | 125.0 | 12.4 | 80.7 | 344.0 | 3299.2 |
| short | 4 | 43 | 451.8 | 38.3 | 26.2 | 97.6 | 10223.6 |
| short | 4 | 43 | 443.8 | 38.4 | 26.1 | 99.3 | 10248.0 |
| short | 4 | 43 | 452.3 | 38.5 | 26.1 | 97.5 | 10266.0 |
| medium | 1 | 568 | 815.7 | 12.8 | 78.5 | 696.4 | 4077.3 |
| medium | 1 | 568 | 823.2 | 12.8 | 78.3 | 690.0 | 4095.6 |
| medium | 1 | 568 | 827.5 | 12.9 | 78.0 | 686.4 | 4110.7 |
| medium | 4 | 568 | 4261.4 | 39.8 | 25.3 | 125.7 | 14398.8 |
| medium | 4 | 568 | 4175.6 | 39.7 | 25.3 | 128.3 | 14308.1 |
| medium | 4 | 568 | 4059.9 | 39.3 | 25.5 | 131.9 | 14093.4 |
| long | 1 | 2513 | 3770.0 | 13.7 | 73.3 | 666.6 | 7265.4 |
| long | 1 | 2513 | 3930.5 | 13.8 | 72.9 | 639.4 | 7445.0 |
| long | 1 | 2513 | 4079.5 | 13.9 | 72.2 | 616.0 | 7628.0 |
| long | 4 | 2513 | 21802.1 | 48.7 | 20.6 | 83.9 | 34215.9 |
| long | 4 | 2513 | 21289.4 | 47.3 | 21.2 | 85.9 | 33356.7 |
| long | 4 | 2513 | 20498.7 | 45.9 | 21.9 | 89.2 | 32193.5 |

### Summary Comparison:
- **Light load (`short`, `medium` conc=1)**: 0% overhead (<0.3% delta, within measurement variance).
- **Medium load (`medium` conc=4)**: TTFT improved from 4165ms to 3850ms (~8% faster).
- **Heavy load (`long` conc=4)**:
  - **TTFT**: Improved from **21.2s** to **14.8s** (**-30% reduction**).
  - **Decode TPS**: Increased from **21.2 tok/s** to **38.8 tok/s** (**+83% faster, nearly 2x throughput**).
  - **E2E Latency**: Dropped from **33.2s** to **21.4s** (**-36% overall latency**).

## Output parity

Output text generation is byte-for-byte identical on greedy sampling. The thread-routing changes only manage dispatch and stream affinity without modifying weights, prompt formatting, tokenization, or model execution.

## Risk notes

- Backwards compatibility: `BatchedEngine._generation_executor` is preserved via property getter/setter so existing consumers (e.g. `ResidencyManager` and external tests) continue to function without modification.
- MLLM compatibility: Multimodal models step directly on the Event Loop thread; `MLXExecutor` does not force thread-hopping for MLLM, preserving existing MLLM execution semantics.

### CODE REVIEW

**Comment by riiji** [comment]
@Thump604 

The basic idea is to isolate MLX executions with some wrapper (I called it MLXExecutor), this wrapper decides how to run mlx (on MT or in a worker)

The problem is that now we sometimes create tensors in MT, sometimes in workers, but we cannot actually do it in both at the same time. I  think you're already know the context of the problem and maybe there was some discussion about it. Looks like it's worth to fix because it gives a very good performance boost
